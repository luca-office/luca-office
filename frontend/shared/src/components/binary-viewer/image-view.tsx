import {css} from "@emotion/react"
import * as React from "react"
import {backgroundColorDarker, CustomStyle} from "../../styles"

interface Props extends CustomStyle {
  readonly url: string
}

export const ImageView: React.FC<Props> = ({customStyles, url}) => (
  <div css={[customStyles, styles.image(url)]} className="image" />
)

const styles = {
  image: (url: string) =>
    css({
      backgroundImage: `url(${url})`,
      backgroundColor: backgroundColorDarker,
      backgroundPosition: "center",
      backgroundRepeat: "no-repeat",
      backgroundSize: "contain"
    })
}
