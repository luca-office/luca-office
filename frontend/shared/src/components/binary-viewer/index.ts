export * from "./binary-viewer"
export * from "./image-view"
export * from "./video-view"
