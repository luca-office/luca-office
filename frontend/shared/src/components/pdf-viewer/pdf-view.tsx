import {css} from "@emotion/react"
import * as React from "react"
import {CustomStyle} from "../../styles"

export interface PdfViewProps extends CustomStyle {
  readonly binaryUrl: string
}

export const PdfView: React.FC<PdfViewProps> = ({binaryUrl, customStyles}) => (
  <div css={[styles.pdf, customStyles]}>
    <object css={styles.object} data={binaryUrl}>
      <a href={binaryUrl}></a>
    </object>
  </div>
)

const styles = {
  pdf: css({
    maxHeight: "100%",
    height: "100%",
    width: "100%",
    maxWidth: "100%",
    display: "flex",
    flexGrow: 1,
    flexDirection: "column"
  }),
  object: css({
    width: "100%",
    flexGrow: 1
  })
}
