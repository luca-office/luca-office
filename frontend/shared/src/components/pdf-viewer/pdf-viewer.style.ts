import {css} from "@emotion/react"
import {
  borderRadius,
  boxHeightSmall,
  cardBottomColor,
  Flex,
  fontColor,
  fontColorLight,
  radius,
  spacing,
  spacingCard,
  spacingLarge,
  spacingSmall,
  spacingTiny,
  zIndex1
} from "../../styles"

const Size = {
  control: 28,
  pagesInput: 57
}

export const pdfViewerStyle = {
  viewer: css(Flex.column),
  subHeader: css({
    zIndex: zIndex1
  }),
  pdfContent: css(Flex.row, {
    justifyContent: "center",
    flexDirection: "column",
    flexGrow: 1,
    overflow: "auto"
  }),
  footerWrapper: css(Flex.row, {
    justifyContent: "center",
    backgroundColor: cardBottomColor,
    height: boxHeightSmall,
    borderRadius: radius(0, 0, borderRadius, borderRadius),
    padding: spacingTiny
  }),
  footer: css(Flex.row),
  footerControls: css(Flex.row, {
    marginRight: spacingLarge + spacingTiny
  }),
  footerControlButtonMarginRight: css({
    marginRight: spacingCard
  }),
  footerPages: css(Flex.row),
  footerPagesInput: css({
    width: Size.pagesInput
  }),
  footerPagesDivider: css({
    margin: spacing(0, spacingSmall)
  })
}
