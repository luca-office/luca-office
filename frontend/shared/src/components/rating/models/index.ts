export * from "./criterion-metadata"
export * from "./rating-coding-criterion"
export * from "./scenario-criterion-selection"
export * from "./table-entity"
