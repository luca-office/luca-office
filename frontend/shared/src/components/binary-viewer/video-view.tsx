import {css} from "@emotion/react"
import React, {useEffect} from "react"
import {VideoConfig} from "../../models"
import {backgroundColorDarker, CustomStyle, Flex} from "../../styles"

interface Props extends CustomStyle, VideoConfig {
  readonly url: string
}

export const VideoView: React.FC<Props> = ({
  customStyles,
  url,
  onPlay,
  onPause,
  onEnded,
  onFullscreenEntered,
  onFullscreenLeft
}) => {
  useEffect(() => {
    const onFullscreenChange = () => {
      if (document.fullscreenElement !== null) {
        onFullscreenEntered?.()
      } else {
        onFullscreenLeft?.()
      }
    }

    document.addEventListener("mozfullscreenchange", onFullscreenChange)
    document.addEventListener("webkitfullscreenchange", onFullscreenChange)
    document.addEventListener("fullscreenchange", onFullscreenChange)

    return () => {
      document.removeEventListener("mozfullscreenchange", onFullscreenChange, false)
      document.removeEventListener("webkitfullscreenchange", onFullscreenChange, false)
      document.removeEventListener("fullscreenchange", onFullscreenChange, false)
    }
  }, [])

  return (
    <div css={[customStyles, styles.videoWrapper]} className="video">
      <video
        css={styles.video}
        src={url}
        role="video"
        controls={true}
        controlsList="nodownload"
        disablePictureInPicture={true}
        onPlay={() => onPlay?.()}
        onPause={() => onPause?.()}
        onEnded={() => onEnded?.()}
      />
    </div>
  )
}

const styles = {
  videoWrapper: css(Flex.column, {
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: backgroundColorDarker,
    minHeight: 0
  }),
  video: css({
    maxWidth: "100%",
    maxHeight: "100%",

    "&:focus": {
      outline: "none"
    }
  })
}
