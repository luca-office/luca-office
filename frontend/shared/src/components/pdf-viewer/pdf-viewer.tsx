import * as React from "react"
import {ViewerToolsType} from "../../enums"
import {Binary} from "../../models"
import {CustomStyle} from "../../styles"
import {Option} from "../../utils"
import {ToolsHeader, ViewerToolsSubHeader} from "../viewer-tools"
import {usePdfViewer} from "./hooks/use-pdf-viewer"
import {PdfView} from "./pdf-view"
import {pdfViewerStyle as styles} from "./pdf-viewer.style"

export interface PdfViewerProps extends CustomStyle {
  readonly binaries: Binary[]
  readonly closeBinary?: (id: UUID) => void
  readonly onClose: () => void
  readonly onMinimize?: () => void
  readonly selectedBinaryId: Option<UUID>
  readonly selectBinaryId?: (id: UUID) => void
}

export const PdfViewer: React.FC<PdfViewerProps> = ({
  onClose,
  onMinimize,
  customStyles,
  binaries,
  closeBinary,
  selectBinaryId,
  selectedBinaryId
}) => {
  const {areControlsDisabled, selectedPdf, onLeftClick, onRightClick, selectPdf} = usePdfViewer(
    binaries,
    selectedBinaryId,
    selectBinaryId,
    closeBinary
  )

  return (
    <div css={[styles.viewer, customStyles]}>
      <ToolsHeader toolType={ViewerToolsType.PDF} onClose={onClose} onMinimizeTool={onMinimize} />
      <ViewerToolsSubHeader
        elements={Option.of(binaries)}
        customStyles={styles.subHeader}
        closeElement={closeBinary}
        selectedElement={selectedPdf}
        selectElement={selectPdf}
        hideNavigationButtons={areControlsDisabled}
        {...{
          ...(!areControlsDisabled && {
            navigateToPrevious: onLeftClick,
            navigateToNext: onRightClick
          })
        }}
      />
      <div css={styles.pdfContent}>
        <PdfView binaryUrl={selectedPdf.map(pdf => pdf.path).getOrElse("")} />
      </div>
      <div css={styles.footerWrapper} />
    </div>
  )
}
