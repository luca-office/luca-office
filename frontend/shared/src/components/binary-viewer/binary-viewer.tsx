import {css} from "@emotion/react"
import * as React from "react"
import {BinaryType, ViewerToolsType} from "../../enums"
import {Binary, VideoConfig} from "../../models"
import {backgroundColorDarker, CustomStyle, Flex, zIndex1} from "../../styles"
import {Option} from "../../utils"
import {PdfView} from ".."
import {ToolsHeader, ViewerToolsSubHeader} from "../viewer-tools"
import {ImageView} from "./image-view"
import {VideoView} from "./video-view"

export interface BinaryViewerProps extends CustomStyle {
  readonly activeBinaryId?: UUID
  readonly binaries: Binary[]
  readonly onCloseBinary?: (id: UUID) => void
  readonly onCloseViewer: () => void
  readonly onMinimize?: () => void
  readonly setActiveBinaryId?: (id: UUID) => void
  readonly type: BinaryType
  readonly videoConfig?: VideoConfig
}

export const BinaryViewer: React.FC<BinaryViewerProps> = ({
  activeBinaryId,
  binaries,
  customStyles,
  onCloseBinary,
  onCloseViewer,
  onMinimize,
  setActiveBinaryId,
  type,
  videoConfig
}) => {
  const areControlsDisabled = binaries.length <= 1 || !setActiveBinaryId
  const selectedBinary = activeBinaryId
    ? Option.of<Binary>(binaries.find(binary => binary.id === activeBinaryId))
    : Option.of(binaries[binaries.length - 1])
  const selectedIndex = selectedBinary.map(binary => binaries.indexOf(binary)).getOrElse(binaries.length - 1)

  const onLeftClick = () => {
    if (setActiveBinaryId && binaries.length) {
      if (selectedIndex === 0) {
        setActiveBinaryId(binaries[binaries.length - 1].id)
      } else {
        setActiveBinaryId(binaries[selectedIndex - 1].id)
      }
    }
  }

  const onRightClick = () => {
    if (setActiveBinaryId && binaries.length) {
      if (selectedIndex === binaries.length - 1) {
        setActiveBinaryId(binaries[0].id)
      } else {
        setActiveBinaryId(binaries[selectedIndex + 1].id)
      }
    }
  }

  const renderBinaryView = (binary: Binary) => {
    switch (type) {
      case BinaryType.Image:
        return <ImageView url={binary.path} customStyles={styles.binary} />
      case BinaryType.Video:
        return (
          <VideoView
            url={binary.path}
            customStyles={styles.binary}
            onPlay={videoConfig?.onPlay}
            onPause={videoConfig?.onPause}
          />
        )
      case BinaryType.PDF:
        return <PdfView binaryUrl={binary.path} customStyles={styles.pdfView} />
      default:
        return null
    }
  }

  return selectedBinary
    .map(binary => (
      <div css={[styles.viewer, customStyles]}>
        <ToolsHeader toolType={getToolTypeByBinaryType(type)} onClose={onCloseViewer} onMinimizeTool={onMinimize} />
        <ViewerToolsSubHeader
          customStyles={styles.subHeader}
          elements={Option.of(binaries)}
          selectedElement={selectedBinary}
          closeElement={onCloseBinary}
          selectElement={setActiveBinaryId}
          navigateToPrevious={areControlsDisabled ? undefined : onLeftClick}
          navigateToNext={areControlsDisabled ? undefined : onRightClick}
        />
        {renderBinaryView(binary)}
      </div>
    ))
    .orNull()
}

const styles = {
  viewer: css(Flex.column),
  subHeader: css({
    zIndex: zIndex1
  }),
  binary: css({
    flex: 1
  }),
  pdfView: css({
    backgroundColor: backgroundColorDarker,
    overflow: "auto"
  })
}

const getToolTypeByBinaryType = (binaryType: BinaryType): ViewerToolsType => {
  switch (binaryType) {
    case BinaryType.Image:
      return ViewerToolsType.Image
    case BinaryType.Video:
      return ViewerToolsType.Video
    case BinaryType.PDF:
      return ViewerToolsType.PDF
    default:
      return ViewerToolsType.File
  }
}
