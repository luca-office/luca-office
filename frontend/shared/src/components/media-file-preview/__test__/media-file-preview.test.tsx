// importing from direct file because of issues of babel loader and spyOn
import {mount} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {MediaFile} from "../../../models"
import {MediaFilePreview} from "../.."
import {ImageView} from "../../binary-viewer"
import {PdfView} from "../../pdf-viewer/pdf-view"
import {fileMockPDF, fileMockPNG} from "../__mocks__/file.mock"

const mediaFile = fileMockPNG as MediaFile

const fileDetail = <MediaFilePreview showInlinePreview onPreviewButtonClick={jest.fn()} file={mediaFile} />

jest.mock("../../spreadsheet/dhx-spreadsheet/dhx-spreadsheet", () => ({
  DhxSpreadsheet: () => <div>mockedSheet</div>
}))

describe("Media File Preview", () => {
  it("renders correctly", async () => {
    const component = create(fileDetail)

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure - image view", async () => {
    const component = mount(fileDetail)
    expect(component.find(ImageView)).toHaveLength(1)
    expect(component.find(PdfView)).toHaveLength(0)
  })

  it("has correct structure - pdf view", async () => {
    const component = mount(
      <MediaFilePreview showInlinePreview onPreviewButtonClick={jest.fn()} file={fileMockPDF as MediaFile} />
    )
    expect(component.find(ImageView)).toHaveLength(0)
    expect(component.find(PdfView)).toHaveLength(1)
  })
})
