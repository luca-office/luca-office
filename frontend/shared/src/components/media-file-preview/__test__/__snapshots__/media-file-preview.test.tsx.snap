// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media File Preview renders correctly 1`] = `
<div
  className="css-1kpejq0-MediaFilePreview"
>
  <div
    className="image css-2uffd0-ImageView"
  />
</div>
`;
